hal/src/common/rtc.rs

```diff
@@ -0,0 +1,305 @@
+    }
+}
+
+// --- Timer / Counter Functionality
+
+impl Periodic for Rtc {}
+impl CountDown for Rtc {
+    type Time = Nanoseconds;
+
+    fn start<T>(&mut self, timeout: T)
+    where
+        T: Into<Self::Time>,
+    {
+        let params = TimerParams::new_us(timeout, self.rtc_clock_freq.0);
+        let divider = params.divider;
+        let cycles = params.cycles;
+
+        // Disable the timer while we reconfigure it
+        self.enable(false);
+
+        // 32 bit counter mode plz
+        self.mode0_ctrla().modify(|_, w| w.mode().count32());
+
+        // Now that we have a clock routed to the peripheral, we
+        // can ask it to perform a reset.
+        self.reset();
+
+        // the SVD erroneously marks swrst as write-only, so we
+        // need to manually read the bit here
+        while self.mode0_ctrla().read().bits() & 1 != 0 {}
+
+        // set cycles to compare to...
+        self.mode0().comp[0].write(|w| unsafe { w.comp().bits(cycles) });
+        self.mode0_ctrla().modify(|_, w| {
+            // set clock divider...
+            match divider {
+                1 => w.prescaler().div1(),
+                2 => w.prescaler().div2(),
+                4 => w.prescaler().div4(),
+                8 => w.prescaler().div8(),
+                16 => w.prescaler().div16(),
+                32 => w.prescaler().div32(),
+                64 => w.prescaler().div64(),
+                128 => w.prescaler().div128(),
+                256 => w.prescaler().div256(),
+                512 => w.prescaler().div512(),
+                1024 => w.prescaler().div1024(),
+                _ => unreachable!(),
+            };
+            // clear timer on match...
+            w.matchclr().set_bit();
+            // and enable RTC.
+            w.enable().set_bit()
+        });
+    }
+
+    fn wait(&mut self) -> nb::Result<(), Void> {
+        if self.mode0().intflag.read().cmp0().bit_is_set() {
+            // Writing a 1 clears the flag
+            self.mode0().intflag.modify(|_, w| w.cmp0().set_bit());
+            Ok(())
+        } else {
+            Err(nb::Error::WouldBlock)
+        }
+    }
+}
+
+impl InterruptDrivenTimer for Rtc {
+    /// Enable the interrupt generation for this hardware timer.
+    /// This method only sets the clock configuration to trigger
+    /// the interrupt; it does not configure the interrupt controller
+    /// or define an interrupt handler.
+    fn enable_interrupt(&mut self) {
+        self.mode0().intenset.write(|w| w.cmp0().set_bit());
+    }
+
+    /// Disables interrupt generation for this hardware timer.
+    /// This method only sets the clock configuration to prevent
+    /// triggering the interrupt; it does not configure the interrupt
+    /// controller.
+    fn disable_interrupt(&mut self) {
+        self.mode0().intenclr.write(|w| w.cmp0().set_bit());
+    }
+}
+
+/// Helper type for computing cycles and divider given frequency
+#[derive(Debug, Clone, Copy)]
+pub struct TimerParams {
```

Comment:
Is this identical copypasta from `timer::TimerParams` ? If so, is there any way we can just use that copy instead of duplicating here?

Reply:
there's a few differences. The timer divider has some gaps that RTC doesn't, and timer is in 16 bit mode, so we need to check if the counter compare value fits in 16 bits whereas here we're 32 bit counter mode. I'll think about this though, there might be a good way to do this.

Reply:
I couldn't think of a clean way to do this, however the timer version of TimerParams is copypasta'd 3 times for SAMD11, SAMD21, and thumbv7em, so that should be definitely cleaned up. I think I'll do that, but not in this PR.

Reply:
#352 unifies TimerParams for the timers